Source/WebKit/NetworkProcess/glib/WebKitOverridingResolver.cpp

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (C) 2022 Igalia S.L.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebKitOverridingResolver.h"
+
+#include <wtf/glib/GUniquePtr.h>
+#include <wtf/glib/WTFGType.h>
+
+using namespace WebKit;
+
+typedef struct {
+    GRefPtr<GResolver> wrappedResolver;
+    HashSet<String> localhostAliases;
+} WebKitOverridingResolverPrivate;
+
+struct _WebKitOverridingResolver {
+    GResolver parentInstance;
+
+    WebKitOverridingResolverPrivate* priv;
+};
+
+struct _WebKitOverridingResolverClass {
+    GResolverClass parentClass;
+};
+
+WEBKIT_DEFINE_TYPE(WebKitOverridingResolver, webkit_overriding_resolver, G_TYPE_RESOLVER)
+
+static GList* createLoobackAddressList()
+{
+    GList* list = nullptr;
+    list = g_list_append(list, g_inet_address_new_loopback(G_SOCKET_FAMILY_IPV4));
+    list = g_list_append(list, g_inet_address_new_loopback(G_SOCKET_FAMILY_IPV6));
```

Comment:
Maybe we could save both addresses in the init and here just return a reference.